### PR TITLE
Rewrite gatsby-link in typescript

### PR DESCRIPTION
**Pull request by herecydev**
## Description

Moves gatsby-link into the typescript world.

## Related Issues
Related to #21995 

## TODO

@blainekasten can you provide some assistance, I have a few things I'm uncertain about:

1. Should the `index.d.ts` be deleted? As I understand this should now be autogenerated, with the typescript files being the source of truth now.
2. Everything is fairly trivial but the ref/innerRef is causing typescript some problems. There's some real confusion in [this definitelytyped thread](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31065) and the type signature is not easy to go through

image

3. Should proptypes be in there still? If so, are they to be autogenerated? Or just a carbon copy of typescript props interface?
4. Babel and lint had quite the fight, around L14 you can my efforts to disable eslint to get babel to transpile correctly

Happy to keep working on it but if you have any suggestions let me know

### CODE REVIEW

File: packages/gatsby-link/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { CSSProperties, Ref } from "react"
+import { NavigateOptions, LinkProps } from "@reach/router"
+
+/* eslint-disable @typescript-eslint/interface-name-prefix */
```
Comment (blainekasten) [inline packages/gatsby-link/src/types.ts]: nit, can you just make this the `eslint-disable-next-line` and put it above `interface Window` ?

File: packages/gatsby-link/src/index.tsx [hunk 2]
```diff
@@ -112,23 +145,23 @@ class GatsbyLink extends React.Component {
         style: { ...this.props.style, ...this.props.activeStyle },
       }
     }
-    return null
+    return {}
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: need to look into this. This might have a functional difference..

File: packages/gatsby-link/src/index.tsx [hunk 3]
```diff
@@ -88,8 +118,11 @@ class GatsbyLink extends React.Component {
     instance.disconnect()
   }
 
-  handleRef(ref) {
-    if (this.props.innerRef && this.props.innerRef.hasOwnProperty(`current`)) {
+  handleRef(ref: HTMLAnchorElement | null): void {
+    if (
+      this.props.innerRef &&
+      Object.prototype.hasOwnProperty.call(this.props.innerRef, `current`)
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: why was this change necessary?

File: packages/gatsby-link/src/index.tsx [hunk 4]
```diff
@@ -62,23 +91,24 @@ class GatsbyLink extends React.Component {
       IOSupported,
     }
     this.handleRef = this.handleRef.bind(this)
+    this.defaultGetProps = this.defaultGetProps.bind(this)
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: tbh, i'd rather run this the other way. Let's change the handleRef and defaultGetProps to static arrow functions

**Comment by blainekasten** [comment]
1. I think we should leave the index.d.ts for now, I need to look into this more
2. The typings on the ref stuff looks fine. Is there a real problem here? sorry if I'm not following.
3. Leave the proptypes in there. We can look into auto-generating them later if we want, but as a library it still provides value to some people.
4. Shouldn't these just be defined on the window interface you have [here](https://github.com/gatsbyjs/gatsby/pull/22027/files#diff-c745a811e20ccc30cff51fb154264f6bR6)?

File: packages/gatsby-link/src/index.tsx (hunk 2 above)
Comment (herecydev) [reply packages/gatsby-link/src/index.tsx]: I used https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/reach__router/index.d.ts#L57 as a reference. The actual code: https://github.com/reach/router/blob/master/src/index.js#L408 simply spreads them. So think there should be no change

File: packages/gatsby-link/src/index.tsx (hunk 3 above)
Comment (herecydev) [reply packages/gatsby-link/src/index.tsx]: https://eslint.org/docs/rules/no-prototype-builtins

File: packages/gatsby-link/src/index.tsx (hunk 4 above)
Comment (herecydev) [reply packages/gatsby-link/src/index.tsx]: Do you mean something like:

```tsx
  static defaultGetProps = <TState,>({ isPartiallyCurrent, isCurrent }: LinkGetProps, props: IGatsbyLinkProps<TState>): IGetProps => {
    if (props.partiallyActive ? isPartiallyCurrent : isCurrent) {
      return {
        className: [props.className, props.activeClassName]
          .filter(Boolean)
          .join(` `),
        style: { ...props.style, ...props.activeStyle },
      }
    }
    return {}
  }
```

File: packages/gatsby-link/src/index.tsx [hunk 5]
```diff
@@ -173,48 +219,53 @@ class GatsbyLink extends React.Component {
             // loaded before continuing.
             navigate(to, { state, replace })
           }
-
-          return true
```
Comment (herecydev) [inline packages/gatsby-link/src/index.tsx]: Double checking this is an appropriate change. As I understand React doesn't respect any return values (typings also want a void method)

**Comment by herecydev** [comment]
>     1. I think we should leave the index.d.ts for now, I need to look into this more
>     3. Leave the proptypes in there. We can look into auto-generating them later if we want, but as a library it still provides value to some people.

These are both restored, I've got some concerns about what the source of truth is. Proptypes, Typescript and the .d.ts all have a different understanding of the props. I can spend some time trying to converge them better, thoughts?

> 2. The typings on the ref stuff looks fine. Is there a real problem here? sorry if I'm not following.

Babel will transpile happily but Typescript compiler is not happy. So the PR isn't mergeable until these are fixed

image

The signature looks something like: `type Ref<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"] | RefObject<T> | null;` so I'm not exactly sure how to fix this.

> 4. Shouldn't these just be defined on the window interface you have here?

This would mean the code has to become `window.__BASE_PATH__` and node won't have access to it.

**Comment by ascorbic** [comment]
We should absolutely aim to move to auto generating the typings. However at the moment we're compiling with Babel, so can't generate them as part of the compilation step. We'd need to either switch to compiling the package with tsc, or run it separately to generate the typings.

File: packages/gatsby-link/src/index.tsx (hunk 4 above)
Comment (blainekasten) [reply packages/gatsby-link/src/index.tsx]: almost. not quite that intense! haha

It just means this:
```js
class Foo {
  // this approach has the `this` bound to the class 
  defaultGetProps = () => {}

  // this approach does not, and requires binding in the constructor
  defaultGetProps() {}
}
```

No need for the `static`

Comment (herecydev) [reply packages/gatsby-link/src/index.tsx]: Of course 🤦‍♂ that's done now

**Comment by Eyas** [comment]
I just saw this PR already attempts this. Sorry for duplicating this work in #22563

Note that I also remove index.d.ts but added a typings generation step to introduce a compatible one. Happy to work on this on either PR
